fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/rules/ImplementRecursiveDfsUnionRule.java

```diff
@@ -0,0 +1,124 @@
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.PlanPartitionMatchers.planPartitions;
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.PlanPartitionMatchers.rollUpPartitions;
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.QuantifierMatchers.forEachQuantifier;
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.QuantifierMatchers.forEachQuantifierOverRef;
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.RecordQueryPlanMatchers.anyPlan;
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.RecordQueryPlanMatchers.mapPlan;
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.RecordQueryPlanMatchers.preOrderTraversalIsAllowed;
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.RecordQueryPlanMatchers.tempTableInsertPlan;
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.RecordQueryPlanMatchers.tempTableScanPlan;
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.ReferenceMatchers.exploratoryMember;
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.RelationalExpressionMatchers.recursiveUnionExpression;
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.RelationalExpressionMatchers.selectExpression;
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.RelationalExpressionMatchers.tempTableInsertExpression;
+import static com.apple.foundationdb.record.query.plan.cascades.matching.structure.SetMatcher.exactlyInAnyOrder;
+
+@SuppressWarnings("PMD.TooManyStaticImports")
+public class ImplementRecursiveDfsUnionRule extends ImplementationCascadesRule<RecursiveUnionExpression> {
+
+    @Nonnull
+    private static final BindingMatcher<RecordQueryPlan> initialInnerPlanMatcher = anyPlan();
+
+    @Nonnull
+    private static final BindingMatcher<TempTableInsertPlan> initialPlanMatcher = tempTableInsertPlan(initialInnerPlanMatcher);
+
+    @Nonnull
+    private static final BindingMatcher<PlanPartition> innerPlanPartitionMatcher = anyPlanPartition();
+
+    @Nonnull
+    private static final BindingMatcher<Collection<PlanPartition>> innerPlanPartitionsMatcher = rollUpPartitions(any(innerPlanPartitionMatcher));
+
+    @Nonnull
+    private static final BindingMatcher<Reference> innerReferenceMatcher = planPartitions(innerPlanPartitionsMatcher);
+
+    @Nonnull
+    private static final BindingMatcher<Quantifier.ForEach> innerQunMatcher = forEachQuantifierOverRef(innerReferenceMatcher);
+
+    @Nonnull
+    private static final BindingMatcher<Quantifier.ForEach> mapTempTableQunMatcher = forEachQuantifier(mapPlan(tempTableScanPlan()));
```

Comment:
As stated below, these matchers should look at the logical side, not the physical implementation. However, it is true that you would still (under the current standings) would have to match both temp table scan expressions as well as simple SELECTs (without predicates) over temp table scan expressions. I filed https://github.com/FoundationDB/fdb-record-layer/issues/3649 to simplify that (and a myriad of other cases). Once that is done, the PRE here should be simplified to reflect that. Please refer to that issue in the code here.